modules/hardware/common/shared-mem.nix

```diff
@@ -0,0 +1,232 @@
+    memSize = mkOption {
+      type = types.int;
+      default = 16;
+      description = mdDoc ''
+        Defines shared memory size in MBytes.
+      '';
+    };
+    hostSocketPath = mkOption {
+      type = types.path;
+      default = "/tmp/ivshmem_socket"; # The value is hardcoded in the application
+      description = mdDoc ''
+        Defines location of the shared memory socket. It's used by qemu
+        instances for memory sharing and sending interrupts.
+      '';
+    };
+    flataddr = mkOption {
+      type = types.str;
+      default = "0x920000000";
+      description = mdDoc ''
+        If set to a non-zero value, it maps the shared memory
+        into this physical address. The value is arbitrary chosen, platform
+        specific, in order not to conflict with other memory areas (e.g. PCI).
+      '';
+    };
+    vms_enabled = mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      description = mdDoc ''
+        If set to a non-zero value, it maps the shared memory
+        into this physical address. The value is arbitrary chosen, platform
+        specific, in order not to conflict with other memory areas (e.g. PCI).
+      '';
+    };
+    enable_host = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = mdDoc ''
+        Enables memsocket on host.
+      '';
+      apply =
+        enable:
+        if enable && !config.ghaf.shm.enable then
+          builtins.throw "Enable shared memory in order to use shm on host"
+        else
+          enable;
+    };
+    instancesCount = mkOption {
+      type = types.int;
+      default =
+        if config.ghaf.shm.enable_host then
+          (builtins.length config.ghaf.shm.vms_enabled) + 1
+        else
+          builtins.length config.ghaf.shm.vms_enabled;
+      description = mdDoc ''
+        Number of VMs entitled to use shared memory.
+      '';
+    };
+    serverSocketPath = mkOption {
+      type = types.path;
+      default = "/run/user/${builtins.toString config.ghaf.users.accounts.uid}/memsocket-server.sock";
+      description = mdDoc ''
+        Defines location of the listening socket.
+        It's used by waypipe as an output socket when running in server mode
+      '';
+    };
+    clientSocketPath = mkOption {
+      type = types.path;
+      default = "/run/user/${builtins.toString config.ghaf.users.accounts.uid}/memsocket-client.sock";
+      description = mdDoc ''
+        Defines location of the output socket. It's fed
+        with data coming from AppVMs.
+        It's used by waypipe as an input socket when running in client mode
+      '';
+    };
+    display = mkOption {
+      type = types.bool;
+      default = false;
+      description = "Display VMs using shared memory";
+      apply =
+        enable:
+        if enable && !config.ghaf.shm.enable then
+          builtins.throw "Enable shared memory in order to use shm display"
+        else
+          enable;
+    };
+  };
```

Comment:
Do you need all of these options to be configurable, or are they referenced outside of the module? If not, you can define them in the `let ... in` part. Also, you can define assertions in the config if needed